distsqlrun: ensure outbox's RowChannel is closed

jordanlewis opened the pull request:
Previously, the outbox could fail during startup without closing its
RowChannel. This could lead to deadlocked flows in rare cases.

Closes #34796 and supplants #34799.

Release note: None

jordanlewis committed:
distsqlrun: ensure outbox's RowChannel is closed

Previously, the outbox could fail during startup without closing its
RowChannel. This could lead to deadlocked flows in rare cases.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35073)
<!-- Reviewable:end -->

andreimatei reviewed (approved):
LGTM

Good find! Is this for sure what happened in #34796 ?


Let's add a outbox unit test. Something like create an outbox but inject an error into the `nodeDialer.Dial` or into `client.FlowStream`. I think it should be doable, although I'm not sure about a precedent.
Maybe we can put the dialer behind some interface, or there's some sort of grpc mock here for example.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35073#-:-LZ6WY3FA4F4edvIh0gC:b2kdn9g)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @asubiotto)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
It's not for sure what happened but it seems extremely likely. If an error happens high in that method and people have already started pushing, nobody will unblock the pushers.

jordanlewis commented:
Added a unit test.

jordanlewis commented:
@andreimatei ptal at the test

andreimatei reviewed (approved):
I don't love the test as written; I've left some comments on it.
But feel free to merge as you see fit.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35073#-:-LZ76GMk65pjvoumdVfP:bs5o5mk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/distsqlrun/outbox_test.go, line 449 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35073#-LZ74n5EAPyGfIU3yxOR:-LZ74n5EAPyGfIU3yxOS:b9711d9) ([raw file](https://github.com/cockroachdb/cockroach/blob/788ad59ed2c857ada9fc80eaed8c8404c3d96a4b/pkg/sql/distsqlrun/outbox_test.go#L449)):*
> ```Go
> // Test that the outbox unblocks its producers if it fails to connect during
> // startup.
> func TestOutboxUnblocksProducers(t *testing.T) {
> ```

what I would test is that the outbox calls `ConsumerClosed` when it fails to connect. The check would be that a pusher gets a closed status from `Push()`, not that anything is unblocked. I think that's a simpler to understand test and less code too.
___
*[pkg/sql/distsqlrun/outbox_test.go, line 453 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35073#-LZ74C4xAh1sAQW_s6hB:-LZ74C4xAh1sAQW_s6hC:b-6s2uk4) ([raw file](https://github.com/cockroachdb/cockroach/blob/788ad59ed2c857ada9fc80eaed8c8404c3d96a4b/pkg/sql/distsqlrun/outbox_test.go#L453)):*
> ```Go
> 
> 	stopper := stop.NewStopper()
> 	defer stopper.Stop(context.TODO())
> ```

nit: declare a `ctx := context.Background()` (or TODO) at the top and use that throughout
___
*[pkg/sql/distsqlrun/outbox_test.go, line 462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35073#-LZ75L6RBRYPiIFEsDnH:-LZ75L6RBRYPiIFEsDnI:b-r9xsf6) ([raw file](https://github.com/cockroachdb/cockroach/blob/788ad59ed2c857ada9fc80eaed8c8404c3d96a4b/pkg/sql/distsqlrun/outbox_test.go#L462)):*
> ```Go
> 		stopper:    stopper,
> 		EvalCtx:    &evalCtx,
> 		nodeDialer: nodedialer.New(newInsecureRPCContext(stopper), staticAddressResolver(nil)),
> ```

I don't really know what's going on here; could you add some comments? What's the behavior when attempting to dial this mysterious `staticNodeID`? It just blocks.
In any case, using context cancelation to get an error from that connection is surprising. If we can inject an error somehow (through a gRPC mock or through some testing knob), that'd seem clear to me.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
TFTR! I cleaned up the mysterious nodeDialer which is not relevant, and added your idea to verify the return value of Push. I left the test to ensure the row channel unblocks its writers despite it being a bit complicated because I think it's important to test this capability somewhere. I also kept context cancellation as the mechanism to cause the error because that's the thing we saw in real life. Adding a testing knob for this seems unnecessary.

andreimatei reviewed (approved):
Still don't like it much, but merge at will.

> I left the test to ensure the row channel unblocks its writers despite it being a bit complicated because I think it's important to test this capability somewhere.

Well then test that in a test that has nothing to do with failure to connect. Also I'd guess we already maybe have such a test, since there's very explicit code to do this unblocking so I'd guess it was introduced with a test.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35073#-:-LZCKhzmCIpHdjo8ZD4Y:bbon4zt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/distsqlrun/outbox_test.go, line 458 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35073#-LZCJlYG7GAYuUp5P61m:-LZCJlYG7GAYuUp5P61n:bbm5qmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ffee429a28722f942308245e213aec8ce9a3b09/pkg/sql/distsqlrun/outbox_test.go#L458)):*
> ```Go
> 	st := cluster.MakeTestingClusterSettings()
> 	evalCtx := tree.MakeTestingEvalContext(st)
> 	defer evalCtx.Stop(context.Background())
> ```

nit: use `ctx` here
___
*[pkg/sql/distsqlrun/outbox_test.go, line 490 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35073#-LZCK6Rt8rt_eCX_anuB:-LZCK6Rt8rt_eCX_anuC:b-v3tegc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ffee429a28722f942308245e213aec8ce9a3b09/pkg/sql/distsqlrun/outbox_test.go#L490)):*
> ```Go
> 
> 	// Cancel the context to force the outbox to fail to connect.
> 	cancel()
> ```

Oh so you're canceling the context *before* we even use some dialer or attempt to connect and that generates some convenient error? This is really brittle and unreadable.
Finding a way to explicitly inject an error would be a lot better?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
Got rid of the confusing `ctx` way of injecting an error, replaced with a `nil` dialer. Going to merge now, TFTR

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35073#-:-LZ_c2MG5o9yXz4jlFqZ:b-1ld4ry)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @asubiotto, and @jordanlewis)

___
*[pkg/sql/distsqlrun/outbox_test.go, line 462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35073#-LZ75L6RBRYPiIFEsDnH:-LZ_c-Ef0PbUD20xjq6P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/788ad59ed2c857ada9fc80eaed8c8404c3d96a4b/pkg/sql/distsqlrun/outbox_test.go#L462)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't really know what's going on here; could you add some comments? What's the behavior when attempting to dial this mysterious `staticNodeID`? It just blocks.
In any case, using context cancelation to get an error from that connection is surprising. If we can inject an error somehow (through a gRPC mock or through some testing knob), that'd seem clear to me.
</blockquote></details>

Done.
___
*[pkg/sql/distsqlrun/outbox_test.go, line 490 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35073#-LZCK6Rt8rt_eCX_anuB:-LZ_bw-u0WcXoIm0jToe:bgowx12) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ffee429a28722f942308245e213aec8ce9a3b09/pkg/sql/distsqlrun/outbox_test.go#L490)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Oh so you're canceling the context *before* we even use some dialer or attempt to connect and that generates some convenient error? This is really brittle and unreadable.
Finding a way to explicitly inject an error would be a lot better?
</blockquote></details>

Got rid of this and passed a `nil` dialer in instead.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
bors r+
